ROCm: Enable OpenMP support always.

kwryankrattiger opened the pull request:
FYI: @eugeneswalker @wspear  This should fix the issues seen buliding `vtk-m +rocm`

#32197

kwryankrattiger committed:
ROCm: ROCm always builds with OpenMP support.

kwryankrattiger commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -94,6 +94,7 @@ class LlvmAmdgpu(CMakePackage):
         ),
     )
     variant("openmp", default=False, description="Enable OpenMP")
+    conflicts("+openmp", msg="llvm-amdgpu OpenMP support is currently disabled")
```
This option is verified to not work with any version `@5:`.

srekolam replied:
@estewart08 , need your help here.

kwryankrattiger commented:
@srekolam I updated this to always build ROCmPackages with OpenMP support. It seems this is what is expected by downstream packages so it makes sense to me to have it always enabled.

srekolam reviewed (commented):
Looks okay to me.

kwryankrattiger commented:
Converting this to a draft. There is a bug in the flang/runtime for 5.1 releases and I am working on a patch for it.

kwryankrattiger commented:
@srekolam There is no 5.2.0 version of the `rocm-openmp-extras` and the 5.1.3 version seems to have a bug in the CMake stuff. It is looking for `<hip-prefix>/../include` in the target but that directory doesn't exist anywhere. 5.2.0 works.

srekolam commented:
> @srekolam There is no 5.2.0 version of the `rocm-openmp-extras` and the 5.1.3 version seems to have a bug in the CMake stuff. It is looking for `<hip-prefix>/../include` in the target but that directory doesn't exist anywhere. 5.2.0 works.

I have submitted the changes for open-mp-extras for 5.2.0 for review - #31667

estewart08 commented:
> The `rocm-dev` package that AMD provides via apt and rpm comes with OpenMP support. Is that OpenMP support coming from `rocm-openmp-extras`? EDIT: Running `apt remove openmp-extras` and trying a few builds reveals that the answer is "yes".

Yes, that is correct.

kwryankrattiger committed:
Convert llvm-amdgpu+openmp -> rocm-openmp-extras

renjithravindrankannath reviewed (dismissed):
Closing the PR [32645](https://github.com/spack/spack/pull/32645) since rocm-openmp-extras will also be added in this PR.

kwryankrattiger committed:
Add handling for external rocm-openmp-extras

llvm-amdgpu is not in the self.spec when using externals for ROCm,
so use the rocm-openmp-extras prefix instead when setting up the
runtime environment.

Note: either both llvm-amdgpu and rocm-openmp-extras are external, or
not. It is not supported to mix with this fix.

kwryankrattiger commented:
@spackbot fix style

eugeneswalker commented:
Any more to be done here? Is this ready for merge?

kwryankrattiger commented:
> Any more to be done here? Is this ready for merge?

It should be ready, just a matter of pipelines passing.

cgmb commented on var/spack/repos/builtin/packages/rocblas/package.py:
```diff
@@ -144,9 +145,27 @@ def check(self):
         "5.2.0",
     ]:
         depends_on("hip@" + ver, when="@" + ver)
-        depends_on("llvm-amdgpu@" + ver, type="build", when="@" + ver)
         depends_on("rocminfo@" + ver, type="build", when="@" + ver)
 
+    for ver in [
+        "3.9.0",
+        "3.10.0",
+        "4.0.0",
+        "4.1.0",
+        "4.2.0",
+        "4.3.0",
+        "4.3.1",
+        "4.5.0",
+        "4.5.2",
+        "5.0.0",
+        "5.0.2",
+        "5.1.0",
+        "5.1.3",
+        "5.2.0",
+    ]:
+        depends_on("llvm-amdgpu@" + ver, type="build", when="@" + ver)
+        depends_on("rocm-openmp-extras@" + ver, type="build", when="@" + ver)
+
```
Why is this change needed? rocBLAS depends on LLVM due to Tensile, but I don't think OpenMP is required for the base rocBLAS library. OpenMP is only used in the tests.

kwryankrattiger replied:
that was a copy-pasta error, I was just pulling it up since the versions needed to match. The new loop is because rocm-openmp-extras doesn't have the full version range.

cgmb commented on var/spack/repos/builtin/packages/rocblas/package.py:
```diff
@@ -110,7 +110,8 @@ class Rocblas(CMakePackage):
     depends_on("googletest@1.10.0:", type="test")
     depends_on("netlib-lapack@3.7.1:", type="test")
-    depends_on("llvm-amdgpu +openmp", type="test")
+    depends_on("llvm-amdgpu", type="test")
+    depends_on("rocm-openmp-extras", type="test")
```
The `depends_on("llvm-amdgpu +openmp", type="test")` was just to add the `+openmp` when building the tests. Is `depends_on("llvm-amdgpu", type="test")` required to make OpenMP work? I would expect `depends_on("rocm-openmp-extras", type="test")` to be sufficient.

kwryankrattiger replied:
That may be the case, but I didn't know all of what was actually needed to make it work so I did the thing the most closely emulated the previous dependency.

cgmb commented on var/spack/repos/builtin/packages/rocfft/package.py:
```diff
@@ -96,7 +96,8 @@ class Rocfft(CMakePackage):
     depends_on("fftw@3.3.8:", type="test")
     depends_on("boost@1.64.0: +program_options", type="test")
-    depends_on("llvm-amdgpu +openmp", type="test")
+    depends_on("llvm-amdgpu", type="test")
+    depends_on("rocm-openmp-extras", type="test")
```
Just like with rocBLAS, `depends_on("llvm-amdgpu +openmp", type="test")` was just to make OpenMP available in the tests. I would expect `depends_on("rocm-openmp-extras", type="test")` to be sufficient. Though, it's possible I'm misunderstanding how OpenMP now works.

cgmb reviewed (dismissed):
I left a couple comments as it wasn't clear to me why the changes to rocBLAS and rocFFT were as large as they were. I suspect that some of the added dependencies are redundant. That said, this change fixes some important issues for OpenMP with ROCm and a few redundant `depends_on` statements won't cause any real harm.

srekolam commented:
@kwryankrattiger .. is there anything waiting for this PR to get merged
?

renjithravindrankannath commented:
Hi @eugeneswalker / @kwryankrattiger ,
Is there any PR in progress to address this issue other than #32151 which is merged.

eugeneswalker commented:
I think, ideally, we would like to be able to build `vtk-m +rocm` WITH `+openmp` unlike right now we are building with `~openmp`. If this PR here is merged, would this let us build `vtk-m +rocm +openmp`? @kwryankrattiger thoughts?

cgmb commented:
Have you tried `spack install --test root rocfft` with this branch? I didn't think it was working, but maybe I'm misremembering. I figured I would try to work around the problem by making OpenMP optional the math library test suites.

srekolam commented:
> Have you tried `spack install --test root rocfft` with this branch? I didn't think it was working, but maybe I'm misremembering. I figured I would try to work around the problem by making OpenMP optional the math library test suites.

Yeah.. I tried that and it was failing.

kwryankrattiger commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -194,7 +194,6 @@ class RocmOpenmpExtras(Package):
         depends_on("hsakmt-roct@" + ver, when="@" + ver)
         depends_on("comgr@" + ver, when="@" + ver)
         depends_on("hsa-rocr-dev@" + ver, when="@" + ver)
-        depends_on("llvm-amdgpu@{0} ~openmp".format(ver), when="@" + ver)
```
This needs to be added back. rocm-openmp-extras is an extension to the llvm-amdgpu package.

kwryankrattiger commented:
@renjithravindrankannath I updated this and fixed the outstanding issues. Would you mind giving this another review?

kwryankrattiger commented:
> @kwryankrattiger .. is there anything waiting for this PR to get merged ?

@srekolam Nothing was immediately waiting, the ROCm stacks that are external have OpenMP enabled. This is only needed by projects that want to build ROCm from source with Spack rather than using externals.

kwryankrattiger committed:
lammps: Remove redundant llvm-amdgpu dependency

eugeneswalker commented:
* [(specs) papi/6nzfnmw 6.0.0.1 gcc@11.1.0 linux-ubuntu20.04-x86_64 E4S](https://gitlab.spack.io/spack/spack/-/jobs/6692894)
```
     1840    openmp.F:5:10:
     1841    
     1842        5 |       use omp_lib
     1843          |          1
  >> 1844    Fatal Error: File '/home/software/spack/__spack_path_placeholder__
             /__spack_path_placeholder__/__spack_path_placeholder__/__spack_pat
             h_placeholder__/__spack_path_placeholder__/__spack_path_placeholde
             r__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_
             path_placeholder__/__spack_path_placeholder__/__spack_path_placeho
             lder__/__spack_path_placeholder__/__spack_path_placeholder__/__spa
             ck_path_placeholder__/__spack_path_placeholder__/__spack_path_plac
             eholder__/__spack_path_placeholder__/__spack_path_placeholder__/__
             spa/linux-ubuntu20.04-x86_64/gcc-11.1.0/rocm-openmp-extras-5.4.3-n
             byzzn3nbzq2l63v7l5yncjobzfgxb62/include/omp_lib.mod' opened at (1)
              is not a GNU Fortran module file
     1845    compilation terminated.
  >> 1846    make[2]: *** [Makefile.recipies:44: openmp] Error 1
```

* [(specs) vtk-m/yiykaza 1.9.0 gcc@11.1.0 linux-ubuntu20.04-x86_64 E4S](https://gitlab.spack.io/spack/spack/-/jobs/6692897)
```
     398    In file included from /tmp/root/spack-stage/spack-stage-vtk-m-1.9.0
            -yiykazac7kpxjsvxctvfme2c7fis3hsk/spack-src/vtkm/cont/openmp/intern
            al/DeviceAdapterAlgorithmOpenMP.h:19:
  >> 399    /tmp/root/spack-stage/spack-stage-vtk-m-1.9.0-yiykazac7kpxjsvxctvfm
            e2c7fis3hsk/spack-src/vtkm/cont/openmp/internal/FunctorsOpenMP.h:25
            :10: fatal error: 'omp.h' file not found
     400    #include <omp.h>
     401             ^~~~~~~
     402    1 error generated when compiling for gfx90a.
  >> 403    make[2]: *** [vtkm/cont/CMakeFiles/vtkm_cont.dir/build.make:874: vt
            km/cont/CMakeFiles/vtkm_cont.dir/Timer.cxx.o] Error 1
```

Any thoughts on the vtk-m one @kwryankrattiger ? 

I am fine disabling `papi+rocm` in E4S CI stack as part of this PR, if needed, except that the error we see there is not something we are seeing without this PR. Not sure the relation of the error to the content of this PR. If the error is arising truly because of the changes in the PR, it would be nice to know what was going on.

kwryankrattiger committed:
Drop patch for now removed ROCm 5.1
